torch/nn/modules/module.py

```diff
@@ -254,32 +257,38 @@ def register_forward_hook(self, hook):
         return handle
 
     def __call__(self, *input, **kwargs):
-        for hook in self._forward_pre_hooks.values():
-            hook(self, input)
-        result = self.forward(*input, **kwargs)
-        for hook in self._forward_hooks.values():
-            hook_result = hook(self, input, result)
-            if hook_result is not None:
-                raise RuntimeError(
-                    "forward hooks should never return any values, but '{}'"
-                    "didn't return None".format(hook))
-        if len(self._backward_hooks) > 0:
-            var = result
-            while not isinstance(var, Variable):
-                if isinstance(var, dict):
-                    var = next((v for v in var.values() if isinstance(v, Variable)))
-                else:
-                    var = var[0]
-            grad_fn = var.grad_fn
-            if grad_fn is not None:
-                for hook in self._backward_hooks.values():
-                    wrapper = functools.partial(hook, self)
-                    functools.update_wrapper(wrapper, hook)
-                    grad_fn.register_hook(wrapper)
+        with torch.jit.scope('%s$%s' % (self.__class__.__name__, self._name),
```

Comment:
I don't think adding this in here is a good idea. It's a quite large context-manager, implemented using `contextlib` and I'm pretty sure it will be causing us performance problems. It will always be iterating over the input, always does the string interpolation, always enters, always goes to C to check Variable flags, etc. This is definitely too much considering that in 99.999% of invocations all this will be unnecessary. We should come up with a way to make this cheap before this is merged.

Also, I'm not a huge fan of the `$` syntax. Maybe let's just do `ClassName<user specified name>` and `ClassName` if nothing was given. This will let you do e.g. `Conv2d_1` and if someone likes `$` they can still add that to their names.

Reply:
I **really** like the trie idea!

As to the cost: would it be ok to pay for it only if we're tracing? We can also reduce the cost by going back to `push_state` and `pop_state` and keeping the context manager for willing users.

OK for the syntax.

Reply:
Yeah, I think paying only when tracing is fine.

Reply:
I removed the context from `__call__` and introduced a `torch.jit._tracing` module-global variable that is `True` when something is tracing and `False` otherwise. This should bring the cost of scopes while not tracing (which is almost always) to ~zero.